lib/block_view_lockups.go

```diff
@@ -0,0 +1,2179 @@
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectUpdateCoinLockupParams")
+	}
+	return nil
+}
+
+//
+// CoinLockupTransfer Transaction Logic
+//
+
+func (bav *UtxoView) _connectCoinLockupTransfer(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error) {
+	var utxoOpsForTxn []*UtxoOperation
+
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight ||
+		blockHeight < bav.Params.ForkHeights.BalanceModelBlockHeight {
+		return 0, 0, nil,
+			errors.Wrapf(RuleErrorLockupTxnBeforeBlockHeight, "_connectCoinLockupTransfer")
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeCoinLockupTransfer {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectCoinLockupTransfer: called with bad TxnType: %s", txn.TxnMeta.GetTxnType().String())
+	}
+
+	// Try connecting the basic transfer without considering transaction metadata.
+	totalInput, totalOutput, utxoOpsForBasicTransfer, err :=
+		bav._connectBasicTransfer(txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(err, "_connectCoinLockupTransfer")
+	}
+	utxoOpsForTxn = append(utxoOpsForTxn, utxoOpsForBasicTransfer...)
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*CoinLockupTransferMetadata)
+
+	// Validate the transfer amount as non-zero.
+	if txMeta.LockedCoinsToTransferBaseUnits.IsZero() {
+		return 0, 0, nil, errors.Wrap(RuleErrorCoinLockupTransferOfAmountZero,
+			"_connectCoinLockupTransfer")
+	}
+
+	// If this is a DeSo lockup, ensure the amount is less than 2**64.
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() && !txMeta.LockedCoinsToTransferBaseUnits.IsUint64() {
+		return 0, 0, nil, errors.Wrap(RuleErrorCoinLockupTransferOfDeSoCausesOverflow,
+			"_connectCoinLockupTransfer")
+	}
+
+	// Validate recipient and profile public keys as valid.
+	var profileEntry *ProfileEntry
+	if len(txMeta.RecipientPublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupTransferInvalidRecipientPubKey, "_connectCoinLockupTransfer")
+	}
+	if len(txMeta.ProfilePublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupTransferInvalidProfilePubKey, "_connectCoinLockupTransfer")
+	}
+	if !txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profileEntry = bav.GetProfileEntryForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profileEntry == nil || profileEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupTransferOnNonExistentProfile, "_connectCoinLockupTransfer")
+		}
+	}
+
+	// Fetch PKIDs for the recipient, sender, and profile.
+	var senderPKID *PKID
+	if _, senderIsParamUpdater :=
+		GetParamUpdaterPublicKeys(blockHeight, bav.Params)[MakePkMapKey(txn.PublicKey)]; senderIsParamUpdater {
+		senderPKID = ZeroPKID.NewPKID()
```

Comment:
wait why do we do this? This seems odd. If the sender is a param updater, we set the sender to the zero PKID? huh? Do you mean to be doing this for ProfilePublicKey? This seems like a bug.